receiver/statsdreceiver/protocol/statsd_parser.go

```diff
@@ -147,63 +144,34 @@ func (p *StatsDParser) Aggregate(line string) error {
 	case statsdGauge:
 		_, ok := p.gauges[parsedMetric.description]
 		if !ok {
-			p.gauges[parsedMetric.description] = buildGaugeMetric(parsedMetric, timeNowFunc())
+			p.gauges[parsedMetric.description] = buildGaugeMetric(parsedMetric, timeNowFunc(), parsedMetric.gaugeValue())
 		} else {
 			if parsedMetric.addition {
 				savedValue := p.gauges[parsedMetric.description].Metrics().At(0).Gauge().DataPoints().At(0).DoubleVal()
-				parsedMetric.floatvalue = parsedMetric.floatvalue + savedValue
-				p.gauges[parsedMetric.description] = buildGaugeMetric(parsedMetric, timeNowFunc())
+				p.gauges[parsedMetric.description] = buildGaugeMetric(parsedMetric, timeNowFunc(), parsedMetric.gaugeValue()+savedValue)
 			} else {
-				p.gauges[parsedMetric.description] = buildGaugeMetric(parsedMetric, timeNowFunc())
+				p.gauges[parsedMetric.description] = buildGaugeMetric(parsedMetric, timeNowFunc(), parsedMetric.gaugeValue())
 			}
 		}
 
 	case statsdCounter:
 		_, ok := p.counters[parsedMetric.description]
 		if !ok {
-			p.counters[parsedMetric.description] = buildCounterMetric(parsedMetric, p.isMonotonicCounter, timeNowFunc())
+			p.counters[parsedMetric.description] = buildCounterMetric(parsedMetric, p.isMonotonicCounter, timeNowFunc(), parsedMetric.counterValue())
 		} else {
 			savedValue := p.counters[parsedMetric.description].Metrics().At(0).Sum().DataPoints().At(0).IntVal()
-			parsedMetric.intvalue = parsedMetric.intvalue + savedValue
-			p.counters[parsedMetric.description] = buildCounterMetric(parsedMetric, p.isMonotonicCounter, timeNowFunc())
+			p.counters[parsedMetric.description] = buildCounterMetric(parsedMetric, p.isMonotonicCounter, timeNowFunc(), parsedMetric.counterValue()+savedValue)
 		}
 
-	case statsdHistogram:
+	case statsdTiming, statsdHistogram:
 		switch p.observeHistogram {
 		case "gauge":
-			p.timersAndDistributions = append(p.timersAndDistributions, buildGaugeMetric(parsedMetric, timeNowFunc()))
-		case "summary":
-			eachSummaryMetric, ok := p.summaries[parsedMetric.description]
-			if !ok {
-				p.summaries[parsedMetric.description] = summaryMetric{
-					name:          parsedMetric.description.name,
-					summaryPoints: []float64{parsedMetric.floatvalue},
-					labelKeys:     parsedMetric.labelKeys,
-					labelValues:   parsedMetric.labelValues,
-					timeNow:       timeNowFunc(),
-				}
-			} else {
-				points := eachSummaryMetric.summaryPoints
-				p.summaries[parsedMetric.description] = summaryMetric{
-					name:          parsedMetric.description.name,
-					summaryPoints: append(points, parsedMetric.floatvalue),
-					labelKeys:     parsedMetric.labelKeys,
-					labelValues:   parsedMetric.labelValues,
-					timeNow:       timeNowFunc(),
-				}
-			}
-		}
-
-	case statsdTiming:
-		switch p.observeTimer {
```

Comment:
it's my understanding from reading the code that the separate case for statsdTiming vs statsdHistogram was to handle the the switch on `observeTimer`, is this no longer needed?

Reply:
ohhh, right. the code was so duplicated I couldn't find where it was different. 
let me fix this.

Reply:
Fixed. This is missing a test, which I will now add.

Reply:
I've added a test that would've broken for the bug I nearly introduced.